dll/win32/shell32/dialogs/dialogs.cpp

```diff
@@ -1063,44 +1073,374 @@ int WINAPI RestartDialogEx(HWND hWndOwner, LPCWSTR lpwstrReason, DWORD uFlags, D
+
+    dwValue = 0;
+    dwSize = sizeof(dwValue);
+    lRet = RegQueryValueExW(hKey,
+                            L"LogonType",
+                            NULL,
+                            &dwType,
+                            (LPBYTE)&dwValue,
+                            &dwSize);
+    RegCloseKey(hKey);
+
+    if (lRet != ERROR_SUCCESS || dwType != REG_DWORD)
+        return FALSE;
+
+    return (dwValue != 0);
+}
+
+static VOID FancyLogoffOnInit(HWND hwnd, PLOGOFF_DLG_CONTEXT pContext)
+{
+    HDC hdc;
+    long lfHeight;
+    
+    hdc = GetDC(NULL);
+    lfHeight = -MulDiv(13, GetDeviceCaps(hdc, LOGPIXELSY), 72);
+    ReleaseDC(NULL, hdc);
+    pContext->hfFont = CreateFontW(lfHeight, 0, 0, 0, FW_MEDIUM, FALSE, 0, 0, 0, 0, 0, 0, 0, L"MS Shell Dlg");
+    SendDlgItemMessage(hwnd, IDC_LOG_OFF_TEXT_STATIC, WM_SETFONT, (WPARAM)pContext->hfFont, TRUE);
+
+    pContext->hImageStrip = LoadBitmapW(shell32_hInstance, MAKEINTRESOURCEW(IDB_IMAGE_STRIP));
+
+    CreateToolTipForButtons(IDC_LOG_OFF_BUTTON, IDS_LOG_OFF_DESC, hwnd, IDS_LOG_OFF_TITLE);
+    CreateToolTipForButtons(IDC_SWITCH_USER_BUTTON, IDS_SWITCH_USER_DESC, hwnd, IDS_SWITCH_USER_TITLE);
+
+    pContext->OldButtonsProc = (WNDPROC)GetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON), GWLP_WNDPROC);
+
+    for (int i = 0; i < NUMBER_OF_BUTTONS; i++)
+    {
+        pContext->bIsButtonHot[i] = FALSE;
+        SetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON + i), GWLP_USERDATA, (LONG_PTR)pContext);
+        SetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON + i), GWLP_WNDPROC, (LONG_PTR)HotButtonSubclass);
+    }
+}
+
 /*************************************************************************
  * LogOffDialogProc
  *
  * NOTES: Used to make the Log Off dialog work
  */
+
 INT_PTR CALLBACK LogOffDialogProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
 {
+    DRAWITEMSTRUCT* pdis = (DRAWITEMSTRUCT*)lParam;
+    PLOGOFF_DLG_CONTEXT pContext;
+    pContext = (PLOGOFF_DLG_CONTEXT)GetWindowLongPtrW(hwnd, GWLP_USERDATA);
+
     switch (uMsg)
     {
         case WM_INITDIALOG:
+        {
+            pContext = (PLOGOFF_DLG_CONTEXT)lParam;
+            SetWindowLongPtrW(hwnd, GWLP_USERDATA, (LONG_PTR)pContext);
+
+            if (pContext->bFriendlyUI)
+                FancyLogoffOnInit(hwnd, pContext);
             return TRUE;
+        }
 
         case WM_CLOSE:
-            EndDialog(hwnd, IDCANCEL);
+            DestroyWindow(hwnd);
+            PostQuitMessage(IDCANCEL);
             break;
 
-#if 0
         case WM_ACTIVATE:
         {
             if (LOWORD(wParam) == WA_INACTIVE)
-                EndDialog(hwnd, 0);
+            {
+                DestroyWindow(hwnd);
+                PostQuitMessage(0);
+            }
             return FALSE;
         }
-#endif
 
         case WM_COMMAND:
             switch (LOWORD(wParam))
             {
+                case IDC_LOG_OFF_BUTTON:
                 case IDOK:
                     ExitWindowsEx(EWX_LOGOFF, 0);
```

Comment:
SE_SHUTDOWN_NAME?

Reply:
@katahiromz What do you mean?

Reply:
SeShutdownPrivilege is acquired elsewhere, as this code is already here.